src/components/TradeWidget/index.tsx

```diff
@@ -64,17 +64,22 @@ const TradeWidget: React.FC = () => {
   const tokens = useMemo(() => tokenListApi.getTokens(fallBackNetworkId), [fallBackNetworkId])
 
   // Listen on manual changes to URL search query
-  const { sell: sellTokenSymbol, receive: receiveTokenSymbol } = useParams()
+  const { sell: sellTokenSymbol, sellAmt, receive: receiveTokenSymbol, receiveAmt } = useParams()
 
   const [sellToken, setSellToken] = useState(
     () => getToken('symbol', sellTokenSymbol, tokens) || getToken('symbol', 'DAI', tokens),
   )
   const [receiveToken, setReceiveToken] = useState(
     () => getToken('symbol', receiveTokenSymbol, tokens) || getToken('symbol', 'USDC', tokens),
   )
+  const sellInputId = 'sellToken'
+  const receiveInputId = 'receiveToken'
+
+  const methods = useForm({ mode: 'onBlur', defaultValues: { [sellInputId]: sellAmt, [receiveInputId]: receiveAmt } })
+  const { handleSubmit, watch, reset } = methods
 
   // Change URL on internal token change
-  useURLParams(`sell=${sellToken.symbol}-0&receive=${receiveToken.symbol}-20`)
+  useURLParams(`sell=${sellToken.symbol}-${watch(sellInputId)}&receive=${receiveToken.symbol}-${watch(receiveInputId)}`)
```

Comment:
ubernit: should we define a couple of constants for `sellAmount` and `receiveAmount` so the line 82 is more declarative.

Also, the standard way of encoding the amounts should be something like:
`?sell=DAI&sell-amount=200&buy=WETH&buy-amount=1`

Reply:
We are currently using a path with slugs. You want to switch to search params?

Reply:
More declarative? Like this?
```ts
  const url = new URLSearchParams({
    sell: `${sellToken.symbol}-${watch(sellInputId)}`,
    receive: `${receiveToken.symbol}-${watch(receiveInputId)}`,
  })

  // Change URL on internal token change
  useURLParams(url.toString())
```

Reply:
@anxolin updated as you suggested, but I don't know.

Feels a bit weird: `/trade/sell=GUSD&sell-amount=50048.29&buy=TUSD&buy-amount=980`

Reply:
I find it ok, I would feel comfortable leaving as `/trade/sell=GUSD&sell-amount=50048.29&buy=TUSD&buy-amount=980`

@DmitryBespalov the part i don't like that much, is having a param with two params on it with our own logic. 

An alternative approach would be to give more entity to the token pair as in many REST APIs, so these URL would be valid:
* `/trade`
* `/trade/GUSD-TUSD`
* `/trade/GUSD-TUSD?sell-amount=50048.29
* `/trade/GUSD-TUSD?sell-amount=50048.29&buy-amount=980`

Would you be more comfortable with this approach?

Reply:
Problem would be, what do we do when the user selects JUST the sell token or buy token?

It's not weird for example sth like?
* /trade/GUSD-none
* /trade/none-TUSD

also, hope that some doesn't create the `NONE` token 

Equally weird to do:
* /trade/GUSD-
* /trade/-TUSD

Or not modifying the URL until the user selected the token...

Reply:
I do not like the idea of GUSD-none, but this is not a case that will ever happen.

We have pre-selected by default DAI-USDC, and the selector only allows valid tokens.

When the user manually types a non-existing token in the url we can just redirect it to the default path.

So if we want to go for this approach, the url will always be fully filled like `/trade/GUSD-TUSD?sell-amount=50048.29&buy-amount=980`

Reply:
Okay, sounds food then to me!

Reply:
I would personally scrap the word `amount` as part of the URL all together
```
/trade/sell-50048.29-TUSD&buy-980-DAI
```

Reply:
Although this idea shortens the url, the initial argument was against too little url params.

Building on top of it, we can shorten the url keeping individual params without losing meaning with:

```
/trade/GUSD-TUSD?sell=4234.43&buy=87312.1
```

What do you guys think of this? @bh2smith @anxolin @DmitryBespalov

Reply:
I think this looks MUCH better! However, if you are worried about uniqueness... you might want to include the user's address/id or `auctionIndex` (if necessary)

Reply:
@bh2smith I don't think we need that.

Anyway, a few hours later, I managed to get the url working, just not quite the way I expected.

Couldn't figure out how to use a `?` in the url like usual url search params :/
The routing just wouldn't work.

So, updated url to:  

```
/trade/TUSD-GUSD/sell=1.0&buy=1.0
```

Reply:
`history.push('?whatever')` should work, no?

Reply:
@bh2smith for encoding params in the URL there's some rules, I think although `/trade/sell-50048.29-TUSD&buy-980-DAI` sounds good in terms of you can almost read it as if it's english, this URL assumes some collocation order of the parameters. Some of them are even optional

I think Leandro proposal is good, cause:
* `/trade`: It's an entry point for doing any trade
* `/trade/TUSD-GUSD`: Allows to trade TUSD for GUSD
* `/trade/TUSD-GUSD?sell=1.0&buy=1.0`: Allows to provide any or both of the optional parameters. Since the token is removed from the query params and becomes part of the document it self, I agree is a good call to drop the word `amount`

Only BUT is that, `sell` and `buy`should be to the right of the question mark imo

Reply:
@alfetopito, maybe this would help: https://github.com/sindresorhus/query-string

Reply:
@Velenir 
> `history.push('?whatever')` should work, no?

The issue is in the route matching. I can push the `?` into history, but the Trade widget is never rendered when I use any `?` in the `path`. Examples:

```
<Route path="/trade/:sell-:buy?sell=:sellAmount&buy=:buyAmount" component={Trade} />
<Route path="/trade/:sell-:buy\?sell=:sellAmount&buy=:buyAmount" component={Trade} />
<Route path="/trade/:sell-:buy\\?sell=:sellAmount&buy=:buyAmount" component={Trade} />
```

From [react-router docs](https://reacttraining.com/react-router/web/api/Route/path-string-string), [path-to-regexp](https://github.com/pillarjs/path-to-regexp/tree/v1.7.0) deals with this.
My guess is that the `?` is being interpreted in their sudo regex rather than as a literal in the url path.

Reply:
Tried again a minimal test:

```tsx
const TestQuestionMark: React.FC = () => {
  const match = useRouteMatch()
  return <div>Matched {match.path}</div>
}

const App: React.FC = () => (
  <>
    <GlobalStyles />
    <Router basename={process.env.BASE_URL}>
      <Layout>
        <Switch>
          <Route path="/test?" component={TestQuestionMark} />
          <Route path="/test\?" component={TestQuestionMark} />
          <Route path="/test\\?" component={TestQuestionMark} />
          <Route path="/test/?" component={TestQuestionMark} />
          <Route path="/test/\?" component={TestQuestionMark} />
          <Route path="/test/\\?" component={TestQuestionMark} />
          <Route component={NotFound} />
        </Switch>
      </Layout>
    </Router>
  </>
)
```

Nothing matches the url `/test?` nor `/test/?`

---

Aaaaand as I was writing this I realized it will never match because that's not part of the url
:man_facepalming:. I need to deal with it as query-params, like Anxo suggested.

Working on it...

Reply:
Yay, now `/trade/TUSD-GUSD?sell=1.0&buy=1.0` works as expected